src/validate_wasm.rs

```diff
@@ -0,0 +1,320 @@
+/// - Markers inserted by the ink! codegen for errors which can't be checked at compile time.
+pub fn validate_import_section(module: &Module) -> Result<()> {
+    let imports = match module.import_section() {
+        Some(section) => section.entries().iter(),
+        None => {
+            // the module does not contain any imports,
+            // hence no further validation is necessary.
+            return Ok(());
+        }
+    };
+    let original_imports_len = imports.len();
+    let mut errs = Vec::new();
+
+    let filtered_imports = imports.filter(|section| {
+        let field = section.field();
+        if field.contains("panic") {
+            errs.push(String::from(
+                "An unexpected panic function import was found in the contract Wasm.\n\
+                This typically goes back to a known bug in the Rust compiler:\n\
+                https://github.com/rust-lang/rust/issues/78744\n\n\
+                As a workaround try to insert `overflow-checks = false` into your `Cargo.toml`.\n\
+                This will disable safe math operations, but unfortunately we are currently not \n\
+                aware of a better workaround until the bug in the compiler is fixed.",
+            ));
+        } else if field.starts_with(INK_ENFORCE_ERR) {
+            errs.push(parse_linker_error(field));
+        }
+
+        match check_import(field) {
+            Ok(_) => true,
+            Err(err) => {
+                errs.push(err);
+                false
+            }
+        }
+    });
+
+    if original_imports_len != filtered_imports.count() {
+        anyhow::bail!(format!(
+            "Validation of the Wasm failed.\n\n\n{}",
+            errs.into_iter()
+                .map(|err| format!("{} {}", "ERROR:".to_string().bold(), err))
+                .collect::<Vec<String>>()
+                .join("\n\n\n")
+        ));
+    }
+    Ok(())
+}
+
+/// Returns `true` if the import is allowed.
+fn check_import(field: &str) -> Result<(), String> {
+    let allowed_prefixes = ["seal", "memory"];
+    if allowed_prefixes
+        .iter()
+        .any(|prefix| field.starts_with(prefix))
+    {
+        Ok(())
+    } else {
+        let msg = format!(
+            "An unexpected import function was found in the contract Wasm: {}.\n\
+            The only allowed import functions are those starting with one of the following prefixes:\n\
+            {}", field, allowed_prefixes.join(", ")
+        );
+        Err(msg)
+    }
+}
+
+/// Extracts the ink! linker error marker from the `field`, parses it, and
+/// returns a human readable error message for it.
+fn parse_linker_error(field: &str) -> String {
+    let encoded = field
+        .strip_prefix(INK_ENFORCE_ERR)
+        .expect("error marker must exist as prefix");
+    let hex = serde_hex::from_hex(&encoded).expect("decoding hex failed");
+    let decoded =
+        <EnforcedErrors as codec::Decode>::decode(&mut &hex[..]).expect("decoding object failed");
```

Comment:
This should probably still return a friendly string explaining that the `EnforcedError` could not be decoded, and the probable cause (a mismatch between the `ink!` version and the local definition).

Reply:
This could occur when e.g. `ink!` adds a new variant, and someone is running an outdated version of `cargo-contract`.